FEXCore/Source/Interface/Core/JIT/Arm64/VectorOps.cpp

```diff
@@ -14,6 +14,766 @@ tags: backend|arm64
+  const auto Op = IROp->C<IR::IROp_VFCMPScalarInsert>();
+  const auto ElementSize = Op->Header.ElementSize;
+
+  LOGMAN_THROW_AA_FMT(ElementSize == 2 || ElementSize == 4 || ElementSize == 8, "Invalid size");
+  const auto SubRegSize = ARMEmitter::ToVectorSizePair(
+    ElementSize == 2 ? ARMEmitter::SubRegSize::i16Bit :
+    ElementSize == 4 ? ARMEmitter::SubRegSize::i32Bit :
+    ARMEmitter::SubRegSize::i64Bit);
+
+  const auto ZeroUpperBits = Op->ZeroUpperBits;
+  const auto Is256Bit = IROp->Size == Core::CPUState::XMM_AVX_REG_SIZE;
+
+  auto ScalarEmitEQ = [this, SubRegSize](ARMEmitter::VRegister Dst, ARMEmitter::VRegister Src1, ARMEmitter::VRegister Src2) {
+    switch (SubRegSize.Scalar) {
+      case ARMEmitter::ScalarRegSize::i16Bit: {
+        fcmeq(Dst.H(), Src1.H(), Src2.H());
+        break;
+      }
+      case ARMEmitter::ScalarRegSize::i32Bit:
+      case ARMEmitter::ScalarRegSize::i64Bit:
+        fcmeq(SubRegSize.Scalar, Dst, Src1, Src2);
+        break;
+      default:
+        break;
+    }
+  };
+  auto ScalarEmitLT = [this, SubRegSize](ARMEmitter::VRegister Dst, ARMEmitter::VRegister Src1, ARMEmitter::VRegister Src2) {
+    switch (SubRegSize.Scalar) {
+      case ARMEmitter::ScalarRegSize::i16Bit: {
+        fcmgt(Dst.H(), Src2.H(), Src1.H());
+        break;
+      }
+      case ARMEmitter::ScalarRegSize::i32Bit:
+      case ARMEmitter::ScalarRegSize::i64Bit:
+        fcmgt(SubRegSize.Scalar, Dst, Src2, Src1);
+        break;
+      default:
+        break;
+    }
+  };
+  auto ScalarEmitLE = [this, SubRegSize](ARMEmitter::VRegister Dst, ARMEmitter::VRegister Src1, ARMEmitter::VRegister Src2) {
+    switch (SubRegSize.Scalar) {
+      case ARMEmitter::ScalarRegSize::i16Bit: {
+        fcmge(Dst.H(), Src2.H(), Src1.H());
+        break;
+      }
+      case ARMEmitter::ScalarRegSize::i32Bit:
+      case ARMEmitter::ScalarRegSize::i64Bit:
+        fcmge(SubRegSize.Scalar, Dst, Src2, Src1);
+        break;
+      default:
+        break;
+    }
+  };
+  auto ScalarEmitUNO = [this, SubRegSize, ZeroUpperBits, Is256Bit](ARMEmitter::VRegister Dst, ARMEmitter::VRegister Src1, ARMEmitter::VRegister Src2) {
+    switch (SubRegSize.Scalar) {
+      case ARMEmitter::ScalarRegSize::i16Bit: {
+        fcmge(VTMP1.H(), Src1.H(), Src2.H());
+        fcmgt(VTMP2.H(), Src2.H(), Src1.H());
+        break;
+      }
+      case ARMEmitter::ScalarRegSize::i32Bit:
+      case ARMEmitter::ScalarRegSize::i64Bit:
+        fcmge(SubRegSize.Scalar, VTMP1, Src1, Src2);
+        fcmgt(SubRegSize.Scalar, VTMP2, Src2, Src1);
+        break;
+      default:
+        break;
+    }
+    // If the destination is a temporary then it is going to do an insert after the operation.
+    // This means this operation can avoid a redundant insert in this case.
+    const bool DstIsTemp = Dst == VTMP1;
+
+    // Combine results and invert directly in VTMP1.
+    orr(VTMP1.D(), VTMP1.D(), VTMP2.D());
+    mvn(ARMEmitter::SubRegSize::i8Bit, VTMP1.D(), VTMP1.D());
```

Comment:
Ideally we would invert the IR op so we can optimize this out if fed to e.g. selects in the future

Reply:
This one is definitely a bit spicy however you slice it right now.
Currently it matches behaviour where the OpcodeDispatcher used to put an MVN.
Theoretically it matches behaviour of ScalarEmitORD, but with a mvn at the end. So we could optimize to ordered.

An additional quirk that gets thrown in here is that SVE adds an unordered fcmp directly in to the ISA, so we could make a more optimal solution in the backend in this case regardless. Since then it turns in to fcmuo + (I think there is an instruction that converts a predicate to a vector mask?)